[Linux] Improve backtracer ELF/DWARF parsing performance.

al45tair opened the pull request:
On Linux we use FileImageSource to read data from ELF images.  We were doing this in a fairly naive manner, which turns out to be a performance issue with larger statically linked binaries on Linux.

Change FileImageSource to use mmap() instead.

rdar://117590155

al45tair committed:
[Linux] Improve backtracer ELF/DWARF parsing performance.

On Linux we use FileImageSource to read data from ELF images.  We were
doing this in a fairly naive manner, which turns out to be a performance
issue with larger statically linked binaries on Linux.

Change FileImageSource to use mmap() instead.

rdar://117590155

al45tair commented:
@swift-ci Please smoke test

mikeash commented on stdlib/public/Backtracing/FileImageSource.swift:
```diff
@@ -20,51 +20,54 @@ import Swift
     }
+    let base = mmap(nil, Int(size), PROT_READ, MAP_FILE|MAP_PRIVATE, _fd, 0)
+    if base == nil || base! == UnsafeRawPointer(bitPattern: -1)! {
+      throw FileImageSourceError.posixError(_swift_get_errno())
```
This error case and the one above will leak `_fd`.

al45tair replied:
Good point. Since we don't actually need `_fd` after calling `mmap()` anyway, I think the thing to do is make it a local variable and use a `defer` to clean it up.

al45tair committed:
[Linux] Don't keep the file descriptor for FileImageSource.

We don't really need it after calling `mmap()`, and not keeping it around
means it's easy to make sure it gets closed properly in all cases.

rdar://117590155

al45tair commented:
@swift-ci Please smoke test

stephentyrone commented on stdlib/public/Backtracing/FileImageSource.swift:
```diff
@@ -20,51 +20,53 @@ import Swift
-      }
-      break
+    let size = buffer.count * MemoryLayout<T>.stride
+    guard Int(addr) < _mapping.count && _mapping.count - Int(addr) >= size else {
```
I don't love the extremely-manual error-prone C-style bounds checks here. Can we make these somewhat cleaner without too much hassle?

stephentyrone replied:
I think this is all something like:
```
let offset = Int(addr)
buffer.withMemoryRebound(to: UInt8.self) { byteBuf in
  guard(_mapping.indices.contains(offset)) else { ... }
  let slice = _mapping[offset ...]
  guard(slice.count >= byteBuf.count) else { ... }
  byteBuf.update(from: slice)
}
```
@glessard might have a better suggestion.

al45tair replied:
Sure, I'm open to suggestions; I don't mind what you wrote above, but I'll give @glessard a chance to chime in in case he's got a better idea.

stephentyrone commented on stdlib/public/Backtracing/FileImageSource.swift:
```diff
@@ -20,51 +20,53 @@ import Swift
+           _mapping.count)
   }
 
   public func fetch<T>(from addr: Address,
```
`addr` is actually an offset, yes? Can we give it a less misleading name?

al45tair replied:
It's really only called `addr` because all of the other implementations of the protocol in question actually genuinely are reading from addresses and not just offsets in a file, but since the name isn't a protocol conformance issue, I don't have any particular problem with changing it…

al45tair replied:
Actually, looking at it, the type still needs to be `Address`, so I don't think we really gain much by renaming it. I think I'm inclined to leave the name alone.

al45tair committed:
[Linux] Improve the bounds checks.

It was pointed out that there was a more Swift-y way to write the
bounds checks for the `fetch<T>` implementation, so do that instead.

rdar://117590155

al45tair commented:
@swift-ci Please smoke test
